### PR TITLE
[No QA] Update to Webpack 5 and peer dependencies 

### PR DESCRIPTION
**Pull request by afmire877**
### Fixed Issues

$ [#10033](https://github.com/Expensify/App/issues/10033)

### Tests

1. Verify you can install app dependencies and that no errors appear. (npm install)
1. Verify you can still run the web locally and use the app normally with no issues. (npm run web)

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps
Marked as No QA because we will know if this worked or not based on:

1. Deploy workflows
1. Regression tests.

There's nothing specific to QA here.

- [ ] Verify that no errors appear in the JS console

#### iOS

Screenshot 2022-08-06 at 03 06 59

### CODE REVIEW

File: src/setup/platformSetup/index.website.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ import 'shim-keyboard-event-key';
 import checkForUpdates from '../../libs/checkForUpdates';
 import Config from '../../CONFIG';
 import DateUtils from '../../libs/DateUtils';
-import {version as currentVersion} from '../../../package.json';
```
Comment (roryabraham) [inline src/setup/platformSetup/index.website.js:9]: Is this change necessary?

File: src/pages/settings/AboutPage/AboutPage.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ import withLocalize, {withLocalizePropTypes} from '../../../components/withLocal
 import withWindowDimensions, {windowDimensionsPropTypes} from '../../../components/withWindowDimensions';
 import MenuItem from '../../../components/MenuItem';
 import Logo from '../../../../assets/images/new-expensify.svg';
-import {version} from '../../../../package.json';
+import pkg from '../../../../package.json';
```
Comment (roryabraham) [inline src/pages/settings/AboutPage/AboutPage.js:16]: Is this change necessary?

File: desktop/start.js [hunk 3]
```diff
@@ -8,7 +8,7 @@ const basePort = 8080;
 portfinder.getPortPromise({
     port: basePort,
 }).then((port) => {
-    const devServer = `webpack-dev-server --config config/webpack/webpack.dev.js --port ${port} --env.platform desktop`;
+    const devServer = `webpack-dev-server --config config/webpack/webpack.dev.js --port ${port} --env platform=desktop`;
```
Comment (roryabraham) [inline desktop/start.js:11]: Is this change necessary?

**Review by roryabraham — CHANGES_REQUESTED** [review]
Overall it looks much better, but I suspect this PR contains a number of unnecessary changes:

1. Getting rid of destructured imports of `package.json` doesn't seem necessary.
1. Replacing `--env.platform=desktop` with `--env platform=desktop` doesn't seem necessary either.

Please review this PR and remove any unnecessary and/or unrelated changes. Thanks!

File: desktop/start.js (hunk 3 above)
Comment (afmire877) [reply desktop/start.js:11]: In Webpack v5, syntax on env variables has changed to be `--env  key=value`


Reference: https://webpack.js.org/guides/environment-variables/#root

File: src/setup/platformSetup/index.website.js (hunk 1 above)
Comment (afmire877) [reply src/setup/platformSetup/index.website.js:9]: Webpack will only allow ESM to have named imports. So non-ESM can only be doing 'default' imports 

[Here is the reference from Webpack 5 docs](https://webpack.js.org/guides/ecma-script-modules/#flagging-modules-as-esm:~:text=Only%20the%20%22default%22%20export%20can%20be%20imported%20from%20non%2DESM.%20Named%20exports%20are%20not%20available)

Screenshot 2022-08-08 at 22 11 44

File: src/pages/settings/AboutPage/AboutPage.js (hunk 2 above)
Comment (afmire877) [reply src/pages/settings/AboutPage/AboutPage.js:16]: Same as https://github.com/Expensify/App/pull/10286?notification_referrer_id=NT_kwDOAgaDu7M0MTQ2ODUxMDIyOjMzOTgxMzcx#discussion_r940681238

File: src/libs/API.js [hunk 4]
```diff
@@ -73,7 +73,7 @@ function makeRequestWithSideEffects(command, apiCommandParameters = {}, onyxData
     // Assemble the data we'll send to the API
     const data = {
         ...apiCommandParameters,
-        appversion: version,
+        appversion: pkg.version,
```
[the remark this answers is not in the document]
Comment (afmire877) [reply src/libs/API.js:76]: Same as https://github.com/Expensify/App/pull/10286?notification_referrer_id=NT_kwDOAgaDu7M0MTQ2ODUxMDIyOjMzOTgxMzcx#discussion_r940681238

File: src/libs/Environment/betaChecker/index.android.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import CONST from '../../../CONST';
-import {version} from '../../../../package.json';
+import pkg from '../../../../package.json';
```
[the remark this answers is not in the document]
Comment (afmire877) [reply src/libs/Environment/betaChecker/index.android.js:2]: Same as https://github.com/Expensify/App/pull/10286?notification_referrer_id=NT_kwDOAgaDu7M0MTQ2ODUxMDIyOjMzOTgxMzcx#discussion_r940681238

File: config/webpack/webpack.common.js [hunk 6]
```diff
@@ -45,6 +45,9 @@ const webpackConfig = ({envFile = '.env', platform = 'web'}) => ({
             filename: 'index.html',
             usePolyfillIO: platform === 'web',
         }),
+        new ProvidePlugin({
+            process: 'process/browser',
```
[the remark this answers is not in the document]
Comment (afmire877) [reply config/webpack/webpack.common.js:49]: `process` has removed the polyfill for this Node.js variable. This variable is still required at runtime by packages like `react-native-reanimated`. 

Here is the error we get.

Screenshot 2022-08-08 at 23 06 20


Doc reference: https://webpack.js.org/migrate/5/#:~:text=process%20is%20not%20defined

**Comment by roryabraham** [comment]
Thanks for the links @afmire877, it's very helpful to understand the context for why those changes were needed. Turns out I was prettymuch wrong on all counts with my previous review 🙃

**Comment by roryabraham** [comment]
You'll need to [retroactively GPG-sign these commits](https://webdevstudios.com/2020/05/26/retroactively-sign-git-commits/) before we'll be able to approve and merge this, however

**Comment by afmire877** [comment]
I have added one more commit to update `@react-navigation/drawer` to fix the warning below.

Screenshot 2022-08-09 at 00 46 13

File: package.json [hunk 7]
```diff
@@ -74,6 +74,7 @@
     "moment": "^2.29.4",
     "moment-timezone": "^0.5.31",
     "onfido-sdk-ui": "^8.1.0",
+    "process": "^0.11.10",
```
[the remark this answers is not in the document]
Comment (afmire877) [reply package.json:77]: Same as https://github.com/Expensify/App/pull/10286#discussion_r940708335

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: package.json [hunk 8]
```diff
@@ -55,7 +55,7 @@
     "@react-native-firebase/crashlytics": "^12.3.0",
     "@react-native-firebase/perf": "^12.3.0",
     "@react-native-picker/picker": "^2.3.1",
-    "@react-navigation/drawer": "6.3.0",
+    "@react-navigation/drawer": "^6.4.3",
```
[the remark this answers is not in the document]
Comment (afmire877) [reply package.json:58]: Nope. The build is not failing so I will revert 👍🏾

**Comment by afmire877** [comment]
Ready for re-review 👀  @parasharrajat @roryabraham

**Comment by afmire877** [comment]
@parasharrajat  👍🏾  done a rebase and npm install

**Comment by roryabraham** [comment]
@afmire877 Sorry, you've got conflicts again.

**Comment by afmire877** [comment]
> @afmire877 Sorry, you've got conflicts again.

Sorted +1

**Comment by roryabraham** [comment]
Hi @afmire877, thanks for updating conflicts again. There's a few more things to address:

1. I'm pretty suspicious of some of the changes in `package-lock.json` – it seems to have gotten out-of-sync somehow. For example, I see that the `octokit` version is changed in this PR even though you didn't change that dependency in `package.json`. That's what's causing the GitHub Actions CI failure you can see above ^. Did you delete `package-lock.json` at some point? I think the easiest way to fix this might be to:
    1. [copy the contents of `package-lock.json` on the upstream `main`](https://raw.githubusercontent.com/Expensify/App/main/package-lock.json)
    1. paste that into your local `package-lock.json`
    1. run `npm install` again
    1. Commit the diff in your `package-lock.json` to this PR
1. Also, please run `npx browserslist@latest --update-db` and commit the diff in the lockfile

**Comment by roryabraham** [comment]
@afmire877 Did you see my last comment? I don't want us to commit the changes to octokit because those are out-of-scope and it's a sign that this lockfile is incorrect.

**Comment by afmire877** [comment]
Yeah just saw after i commeted. I will address it now

**Comment by afmire877** [comment]
Done @roryabraham ✅

**Comment by roryabraham** [comment]
Very sorry, there are conflicts again. I'll try to review this one faster.

**Comment by afmire877** [comment]
@roryabraham no problem. Rebased again

**Comment by roryabraham** [comment]
Looks like we've got conflicts, lint errors, and unit test failures

**Comment by roryabraham** [comment]
@afmire877 I'm not sure how it's related to this pull request, but I'm now seeing some new console errors:

image

**Comment by roryabraham** [comment]
Maybe we can try this?

```
resolve: { fallback: { 'process/browser': require.resolve('process/browser'), } }
```

**Comment by roryabraham** [comment]
We could maybe also do [this solution](https://stackoverflow.com/questions/69427025/programmatic-webpack-jest-esm-cant-resolve-module-without-js-file-exten), but I think the first one looks preferable.

**Comment by roryabraham** [comment]
Seems like progress has stalled here – @afmire877 when can we expect an update?

**Comment by afmire877** [comment]
Hey, @roryabraham I will get to you in the next few days. I apologise for the late replies.

**Review by roryabraham — APPROVED** [review]
- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Additional review notes
- There is a console warning that pops up, but it is unrelated and can be handled in a separate PR. You will notice that it takes over the screen (which it did not before), but I'm hopeful that will incentivize us to fix console warning more quickly in the future.
- There's some caching stuff I'd like to follow-up on to make webpack builds faster.
- Also testing a production build with `npm run build` and it seemed to work.

**Comment by roryabraham** [comment]
Tests were passing
